src/version.rs

```diff
@@ -36,8 +38,50 @@ impl fmt::Display for HttpVersion {
     }
 }
 
+impl FromStr for HttpVersion {
+    type Err = Error;
+    fn from_str(s: &str) -> Result<HttpVersion, Error> {
+        Ok(match s {
+            "HTTP/0.9" => Http09,
+            "HTTP/1.0" => Http10,
+            "HTTP/1.1" => Http11,
+            "h2" => H2,
+            "h2c" => H2c,
+            _ => return Err(Error::Version),
+        })
+    }
+}
+
 impl Default for HttpVersion {
     fn default() -> HttpVersion {
         Http11
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use std::str::FromStr;
+    use super::HttpVersion;
+    use super::HttpVersion::{Http09,Http10,Http11,H2,H2c};
+
+    #[test]
+    fn test_default() {
+        assert_eq!(Http11, HttpVersion::default());
+    }
+
+    #[test]
+    fn test_from_str() {
+        assert_eq!(Http09, HttpVersion::from_str("HTTP/0.9").unwrap());
+        assert_eq!(Http10, HttpVersion::from_str("HTTP/1.0").unwrap());
+        assert_eq!(Http11, HttpVersion::from_str("HTTP/1.1").unwrap());
+        assert_eq!(H2, HttpVersion::from_str("h2").unwrap());
+        assert_eq!(H2c, HttpVersion::from_str("h2c").unwrap());
+    }
+
+    #[test]
+    #[should_panic]
+    fn test_from_str_panic() {
+        HttpVersion::from_str("foo").unwrap();
```

Comment:
Prefer a test against the `Error::Version` return value instead. This kind of test doesn't notice if the panic came from inside `from_str`, or from the `unwrap`.